qiskit/algorithms/phase_estimators/iphase_estimation.py

```diff
@@ -0,0 +1,193 @@
+        for k in range(problem['num_iterations'], 0, -1):
+            omega_coef /= 2
+            if self._quantum_instance.is_statevector:
+                qc = self.construct_circuit(problem, k, -2 * numpy.pi * omega_coef,
+                                            measurement=False)
+                result = self._quantum_instance.execute(qc)
+                complete_state_vec = result.get_statevector(qc)
+                ancilla_density_mat = qiskit.quantum_info.partial_trace(
+                    complete_state_vec,
+                    range(problem['unitary'].num_qubits)
+                )
+                ancilla_density_mat_diag = numpy.diag(ancilla_density_mat)
+                max_amplitude = max(ancilla_density_mat_diag.min(),
+                                    ancilla_density_mat_diag.max(), key=abs)
+                x = numpy.where(ancilla_density_mat_diag == max_amplitude)[0][0]
+            else:
+                qc = self.construct_circuit(problem, k, -2 * numpy.pi * omega_coef,
+                                            measurement=True)
+                measurements = self._quantum_instance.execute(qc).get_counts(qc)
+
+                if '0' not in measurements:
+                    if '1' in measurements:
+                        x = 1
+                    else:
+                        raise RuntimeError('Unexpected measurement {}.'.format(measurements))
+                else:
+                    if '1' not in measurements:
+                        x = 0
+                    else:
+                        x = 1 if measurements['1'] > measurements['0'] else 0
+            omega_coef = omega_coef + x / 2
+        return omega_coef
+
+    def estimate(self,
+                 num_iterations: int,
+                 unitary: QuantumCircuit,
+                 state_preparation: QuantumCircuit
+                 ) -> 'IPhaseEstimationResult':
+        """
+        Estimate the eigenphase of the input unitary and initial-state pair.
+
+        Args:
+            num_iterations: The number of iterations (rounds) of the phase estimation to run.
+            unitary: The circuit representing the unitary operator whose eigenvalue (via phase)
+                     will be measured.
+            state_preparation: The circuit that prepares the state whose eigenphase will be
+                     measured.  If this parameter is omitted, no preparation circuit
+                     will be run and input state will be the all-zero state in the
+                     computational basis.
+
+        Returns:
+            Estimated phase in an IPhaseEstimationResult object.
+        """
+
+        problem = {'num_iterations': num_iterations,
+                   'unitary': unitary,
+                   'state_preparation': state_preparation}
+        phase = self._estimate_phase_iteratively(problem)
+
+        return IPhaseEstimationResult(num_iterations, phase)
+
+
+class IPhaseEstimationResult(AlgorithmResult):
```

Comment:
Should this derive from  the `PhaseEstimatorResult` (the interface, not PhaseEstimation's result class)?

Reply:
The most important method of `PhaseEstimatorResult` is `most_likely_phase`. The most important method of `IterativePhaseEstimationResult` is `phase`. In the latter case, there is only one phase to return. Both methods return one phase, but the meaning of the phase returned is different. I suppose you could change the `PhaseEstimatorResult` method to `phase`, with the understanding that you are returning a single "best" phase, and it  is up to the implementation to decide how "best" is determined. In one case it would be the most likely phase. In the other, it would be the only phase.

Reply:
I see the point, however I think we should be able to use PE and IPE interchangeably, since they both solve the same problem. Can we find a name that is suitable for both? Maybe call both just `phase` and be clear in the docstring how that phase is determined?

Reply:
Done